iroh-gossip/src/proto/hyparview.rs

```diff
@@ -0,0 +1,707 @@
+                .unwrap();
+
+            self.refill_active_from_passive(&[&removed_peer], io);
+
+            removed_peer
+        })
+    }
+
+    fn refill_active_from_passive(&mut self, skip_peers: &[&PA], io: &mut impl IO<PA>) {
+        if self.active_view.len() + self.pending_neighbor_requests.len()
+            >= self.config.active_view_capacity
+        {
+            return;
+        }
+        // "When a node p suspects that one of the nodes present in its active view has failed
+        // (by either disconnecting or blocking), it selects a random node q from its passive view and
+        // attempts to establish a TCP connection with q. If the connection fails to establish,
+        // node q is considered failed and removed from p’s passive view; another node q′ is selected
+        // at random and a new attempt is made. The procedure is repeated until a connection is established
+        // with success." (p7)
+        let mut skip_peers = skip_peers.to_vec();
+        skip_peers.extend(self.pending_neighbor_requests.iter());
+
+        if let Some(node) = self
+            .passive_view
+            .pick_random_without(&skip_peers, &mut self.rng)
+        {
+            let priority = match self.active_view.is_empty() {
+                true => Priority::High,
+                false => Priority::Low,
+            };
+            let message = Message::Neighbor(Neighbor {
+                priority,
+                data: self.me_data.clone(),
+            });
+            io.push(OutEvent::SendMessage(*node, message));
+            // schedule a timer that checks if the node replied with a neighbor message,
+            // otherwise try again with another passive node.
+            io.push(OutEvent::ScheduleTimer(
+                self.config.neighbor_request_timeout,
+                Timer::PendingNeighborRequest(*node),
+            ));
+            self.pending_neighbor_requests.insert(*node);
+        };
+    }
+
+    fn handle_pending_neighbor_timer(&mut self, peer: PA, io: &mut impl IO<PA>) {
+        if self.pending_neighbor_requests.remove(&peer) {
+            self.passive_view.remove(&peer);
+            self.refill_active_from_passive(&[], io);
+        }
+    }
+
+    fn remove_active_by_index(
+        &mut self,
+        peer_index: usize,
+        respond: Respond,
+        reason: RemovalReason,
+        io: &mut impl IO<PA>,
+    ) -> Option<PA> {
+        if let Some(peer) = self.active_view.remove_index(peer_index) {
+            if respond {
+                let message = Message::Disconnect(Disconnect {
+                    alive: true,
+                    respond: false,
+                });
+                io.push(OutEvent::SendMessage(peer, message));
+            }
+            io.push(OutEvent::DisconnectPeer(peer));
+            io.push(OutEvent::EmitEvent(Event::NeighborDown(peer)));
+            let data = self.peer_data.remove(&peer).unwrap();
```

Comment:
I agree we want to ensure active peer data is present which is likely true right now, but an expect would be appreciated if this explodes
```suggestion
            let data = self.peer_data.remove(&peer).expect("data for peer in active view is known");
```